pandas/core/common.py

```diff
@@ -2439,6 +2439,8 @@ def _get_dtype_type(arr_or_dtype):
         return np.dtype(arr_or_dtype).type
     elif isinstance(arr_or_dtype, CategoricalDtype):
         return CategoricalDtypeType
+    elif arr_or_dtype is NotImplemented:
+        return type(NotImplemented)
```

Comment:
Based purely on the name, function expects a numpy/pandas array or dtype object, and `NotImplemented` is certainly not one of those :). Returning arbitrary output instead of raising an error feels like an anti-pattern to me.

In contrast, for something like `is_bool_dtype` it does make sense to return `False` when given `NotImplemented`.